### PR TITLE
feature/ch51186/usb-cloud-connection-status

### PR DESCRIPTION
**Pull request by busticated**
## Description

Adds `particle usb cloud-status <deviceID>` command to query the device via USB and return its cloud connection status: `unknown`, `connecting`, `connected`, `disconnecting`, `disconnected` ([ch](https://app.clubhouse.io/particle/story/51186/add-particle-usb-connection-state-to-report-current-connection-state-disconnected-connected-etc))

## How to Test

run `particle usb cloud-status --help` and try out the command and options.

## Related Issues / Discussions

https://github.com/particle-iot/particle-usb/pull/24

## Completeness

- [x] User is totes amazing for contributing!
- [x] Contributor has signed [CLA](https://docs.google.com/a/particle.io/forms/d/1_2P-vRKGUFg5bmpcKLHO_qNZWGi5HKYnfrrkd-sbZoA/viewform)
- [x] Problem and solution clearly stated
- [x] Tests have been provided
- [x] Docs have been updated
- [x] CI is passing

### CODE REVIEW

File: src/cli/usb.js [hunk 1]
```diff
@@ -125,6 +125,29 @@ module.exports = ({ commandProcessor, root }) => {
...
+			'until': {
+				description: 'Poll your device for a specific connection state and then exit',
+				choices: [
+					'unknown',
+					'disconnected',
+					'connecting',
+					'connected',
+					'disconnecting'
+				]
+			}
```
Comment (busticated) [inline src/cli/usb.js:140]: 🤔 add optional `--timeout` flag and enforce a timeout when `--until` flag is set..?

File: src/cmd/usb.js [hunk 2]
```diff
@@ -169,6 +169,43 @@ module.exports = class UsbCommand {
...
+			.then(deviceMgr => deviceMgr.status())
+			.then(status => {
+				if (until && until !== status){
+					throw new Error(`Unexpected status: ${status}`);
+				}
+				console.log(status);
+			})
+			.catch(error => {
+				if (until){
+					return this.cloudStatus(args);
```
Comment (busticated) [inline src/cmd/usb.js:202]: 🤪going to need a default timeout too

Comment (busticated) [reply src/cmd/usb.js]: ✅ in https://github.com/particle-iot/particle-cli/pull/577/commits/8977b22cfb8f1973681402dbcbc308dbc24342b4

**Comment by busticated** [comment]
@sergeuz i don't think it's related to my changes as this isn't new behavior (though perhaps my updates make it worse) but i get `IN control transfer failed: LIBUSB_TRANSFER_TIMED_OUT` from time to time as well as `IN control transfer failed: LIBUSB_TRANSFER_STALL` - any ideas what might be going on?

**Comment by busticated** [comment]
> i get IN control transfer failed: LIBUSB_TRANSFER_TIMED_OUT from time to time

i can reproduce this 100% of the time by connecting my device, letting my computer go to sleep, waking it, and running any `usb` command. unplugging and plugging back in gets things back to working.

**Comment by busticated** [comment]
@sergeuz can you think of a way to trigger "disconnecting" and / or "disconnected" states in these `e2e` tests? i haven't been able to observe a device reporting "disconnecting" for example.

**Comment by sergeuz** [comment]
> can you think of a way to trigger "disconnecting" and / or "disconnected" states in these e2e tests? i haven't been able to observe a device reporting "disconnecting" for example.

It can be due to a transition from "disconnecting" to "disconnected" happening almost immediately, or do you mean you don't ever get "disconnected" as well?

**Comment by busticated** [comment]
@sergeuz 

> do you mean you don't ever get "disconnected" as well?

correct - i never see "disconnecting" or "disconnected" just "unknown", "connecting", and "connected"

**Comment by sergeuz** [comment]
> i don't think it's related to my changes as this isn't new behavior (though perhaps my updates make it worse) but i get IN control transfer failed: LIBUSB_TRANSFER_TIMED_OUT from time to time as well as IN control transfer failed: LIBUSB_TRANSFER_STALL - any ideas what might be going on?

No idea, unfortunately. There seems to be some issue in the USB HAL implementation and, possibly, particle-usb as well. @technobly Is this something your team is currently looking into?

**Comment by sergeuz** [comment]
> i never see "disconnecting" or "disconnected"

@busticated Even if the device is indeed disconnected?

**Comment by busticated** [comment]
@sergeuz 

> Even if the device is indeed disconnected?

yes? i tried putting the device into listening mode and then querying but in that case it returns `unknown`. polling after issuing a reset reports `unknown` -> `connecting` -> `connected`

**Comment by sergeuz** [comment]
> i tried putting the device into listening mode and then querying but in that case it returns unknown. polling after issuing a reset reports unknown -> connecting -> connected

That's weird. I'm not sure why "unknown" is reported, it looks like a parsing error somewhere in `fromProtobufEnum()`. I'll take a look.

File: src/cmd/usb.js [hunk 3]
```diff
@@ -169,6 +169,54 @@ module.exports = class UsbCommand {
...
+				let getStatus = usbDevice
+					? usbDevice.getCloudConnectionStatus()
+					: Promise.resolve('unknown');
+
+				return getStatus.then(status => status.toLowerCase());
+			}
+		};
+
+		const options = { id: deviceID, api: this._api, auth: this._auth };
+		const queryDevice = openUsbDeviceById(options)
```
[the remark this answers is not in the document]
Comment (busticated) [reply src/cmd/usb.js:200]: added a `TODO` in https://github.com/particle-iot/particle-cli/commit/ae6fb5da0bd004c0e11cdd625808575323267564#diff-e14fdb0de89920de351a0eb0179e6f9cR44 🤗

File: src/cmd/usb.js [hunk 4]
```diff
@@ -169,6 +169,54 @@ module.exports = class UsbCommand {
...
+		return spin(queryDevice, 'Querying device...')
+			.then(status => {
+				if (until && until !== status){
+					throw new Error(`Unexpected status: ${status}`);
+				}
+				console.log(status);
+			})
+			.catch(error => {
+				if (Date.now() - (started + timeout) > 0){
+					throw new Error('timed-out waiting for status...');
```
[the remark this answers is not in the document]
Comment (busticated) [reply src/cmd/usb.js]: ✅ in https://github.com/particle-iot/particle-cli/pull/577/commits/ae6fb5da0bd004c0e11cdd625808575323267564

File: src/cli/usb.js (hunk 1 above)
Comment (busticated) [reply src/cli/usb.js:145]: ✅ in https://github.com/particle-iot/particle-cli/pull/577/commits/94b64b3051ef95a79cd3224d2879f66dc3477e15

**Comment by busticated** [comment]
self-review :shipit:

**Pull request merged**